ls1intum/Artemis
# [Feature] Add impact warning on feedback blocks for textual exercises

### Checklist

### Motivation and Context


This feature allows text exercise graders (tutors) to view the impact of their grading on the other submissions. We expect that having the graders know more about how their added feedback might affect other submissions in terms of how many automatic feedbacks are generated, will make the grading process more qualitative and also more efficient. That's because graders can become more conscious about the effects of their grading and provide better feedback but also also as they are grading similar submissions they will encounter more of their own feedback, which means reducing their own grading time as well.

### Description

On the server side:
- A field named `numberOfAffectedSubmissions` is added to the **TextBlock** entity as a non-persistent value. The end goal is to calculate and set this value on the fly and return it to the client since adding a new column was seen to be inefficient. 
- A new query function `getNumberOfOtherBlocksInCluster` is added on the **TextBlockRepository**. The query takes the block id as an input, finds its respective cluster and counts and returns how many other blocks exist in that cluster. As elements in a cluster represent blocks which have a high degree of similarity between each other (=> distanceBetweenBlocks < 1 = DISTANCE_THRESHOLD) the number of elements in a block should represent an approximate number of how many other submissions would receive an AUTOMATIC Feedback label - if the feedback would be submitted. 
-  A new function : `setNumberOfAffectedSubmissionsPerBlock` is added on the **TextBlockService** which:
     - iterates over the blocks of a given submission,
     - for each block uses the query `getNumberOfOtherBlocksInCluster` to find and set the block's `numberOfAffectedSubmissions` field value.
- A check is added on the **prepareSubmissionForAssessment** function which checks whether Automatic Feedback Suggestions are enabled and then calls **setNumberOfAffectedSubmissionsPerBlock** on the referenced submission being prepared. The value is thereafter added to the the TextBlock entity which is then returned to the client through the _/submission/{submissionId}_ endpoint 

On the client side: 
- A card-header element with a warning icon and a text is added on **textblock-feedback-editor.component.html**. The text contains a translated warning sentence with a variable number.  This number is directly fetched from the received TextBlock's `numberOfAffectedSubmissions` field. In case this number is 0 or the feedback has 0 credits then the warning is not shown.
- A field `numberOfAffectedSubmissions` is added on **text-block.model.ts** mirroring the same added field on the server side. 
- Translations for the warning sentence are also added in the respective .json files


### Steps for Testing


_Prerequisite:_ 
If you are running locally, set up and run **Athene** on you machine using Docker. 

**1.** Log in to Artemis
**2.** Create a **TextExercise** and make sure to enable **Automatic assessment suggestions**. Also accordingly set the release, due date and assessment due date. Due date can be some day in the future (You will edit this in further steps)
**3.** Add at least 11 students to the course for this exercise and make a simple submission from each of those students account before the deadline
    - An example submission can be just a text: `submission` for all the students. There is surely a large list of permutations but this is one edge case for this feature, and more simple to test. 
**4.** After submitting >=11 submissions from 11 students log in as an admin/tutor and edit the TextExercise Deadline. Set it to ~2 minutes in the future and save the changes. In 2 minutes Artemis should send jobs to Athene for segmentation, embedding and clustering based on the submissions you entered. The results are then sent back to Artemis and written onto the Database where the **text_block** table is updated with the retrieved TextBlock information. For context only, this is what the **text_block** table usually looks like after the deadline is over:
![image](https://user-images.githubusercontent.com/18335489/118555938-925ac680-b763-11eb-882e-36f48889dd75.png)
**5.** Now go to the TextExercise you created and pick a submission to grade. Select a text block as shown below and increase the number of points: 
![affected_submissions_view](https://user-images.githubusercontent.com/18335489/118558905-8b35b780-b767-11eb-82f7-0f2ed00826dd.gif)

**6.** The warning shown in the feedback editor **should** reflect the number of other textblocks in the same cluster. As you might not be able to see the clusters/database on the test server, submitting 11 identical submissions would be the way to go as the resulting number should be 10 in that case.


### Test Coverage




TextAssessmentIntegrationTest.java : 
textblock-feedback-editor.component.ts : 

### Screenshots

![image](https://user-images.githubusercontent.com/18335489/118553787-13648e80-b761-11eb-9ce6-440cdb1e9b51.png)
